lib/resty/http.lua

```diff
@@ -791,38 +817,53 @@ function _M.read_response(self, params)
     if ok then
         if (version == 1.1 and str_find(connection, "close", 1, true)) or
            (version == 1.0 and not str_find(connection, "keep-alive", 1, true)) then
-            self.keepalive = false
+            self.keepalive_supported = false
         end
     else
         -- no connection header
         if version == 1.0 then
-            self.keepalive = false
+            self.keepalive_supported = false
         end
     end
 
     local body_reader = _no_body_reader
-    local trailer_reader, err
+    local trailer_reader
     local has_body = false
+    local has_trailer = false
+    -- If there are no trailers - fully reading response body means socket is ready to be pooled
+    local body_reader_keepalive_ready_callback = _reader_keepalive_ready_mark(self)
+
+    if res_headers["Trailer"] then
+        has_trailer = true
+        -- If there are trailers - fully reading response body doesn't mean socket is ready to be pooled
+        body_reader_keepalive_ready_callback = _reader_keepalive_ready_no_op()
+    end
```

Comment:
Here we call both callbacks if there are trailers (which creates two extra coroutines instead of one). Less of an issue if we can ditch the coroutines, but closures aren't free either, so can this be an if / else instead?